File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ans = INT_MIN, prevMax = 1, prevMin = 1;
        
        for(int i = 0; i < nums.size(); i++) {
            int currMax = max(nums[i], max(prevMax * nums[i], prevMin * nums[i]));
            int currMin = min(nums[i], min(prevMax * nums[i], prevMin * nums[i]));
            prevMax = currMax;
            prevMin = currMin;
            ans = max(ans, prevMax);
        }
        
        return ans;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let result = Number.NEGATIVE_INFINITY;
    let prevMax = 1;
    let prevMin = 1;

    for(let i = 0; i < nums.length; i++) {
        const currMax = Math.max(nums[i], prevMax * nums[i], prevMin * nums[i]);
        const currMin = Math.min(nums[i], prevMax * nums[i], prevMin * nums[i]);
        prevMax = currMax;
        prevMin = currMin;

        result = Math.max(result, currMax)
    }

    return result;
};